cpp/splinepy/py/py_multi_patch.hpp

```diff
@@ -0,0 +1,2412 @@
+/// @param dim_if_none
+/// @param nthreads
+/// @return
+inline std::vector<PySpline::CoreSpline_>
+ToCoreSplineVector(py::list pysplines,
+                   const int para_dim_if_none,
+                   const int dim_if_none,
+                   const int nthreads) {
+  // prepare return obj
+  const int n_splines = static_cast<int>(pysplines.size());
+  std::vector<PySpline::CoreSpline_> core_splines(n_splines);
+
+  auto to_core = [&](int begin, int end) {
+    for (int i{begin}; i < end; ++i) {
+      // get accessor
+      auto spline = pysplines[i];
+
+      // if None, fill null spline
+      if (spline.is_none()) {
+        core_splines[i] =
+            splinepy::splines::kNullSplineLookup[para_dim_if_none - 1]
+                                                [dim_if_none - 1];
+        continue;
+      }
+
+      // not none,
+      core_splines[i] =
+          spline.template cast<std::shared_ptr<PySpline>>()->Core();
+    }
+  };
+  splinepy::utils::NThreadExecution(to_core, n_splines, nthreads);
+
+  return core_splines;
+}
+
+/// @brief
+/// @param splist
+/// @param nthreads
+/// @return
+inline py::list ToPySplineList(CoreSplineVector& splist) {
+  // prepare return obj
+  const int n_splines = static_cast<int>(splist.size());
+
+  // to return
+  py::list pyspline_list(n_splines);
+
+  // cast - single threads, as we should respect GIL
+  for (int i{}; i < n_splines; ++i) {
+    pyspline_list[i] = std::make_shared<PySpline>(splist[i])->ToDerived();
+  }
+
+  return pyspline_list;
+}
+
+/// @brief raises if there's any mismatch between specified properties and all
+/// the entries in the vector.
+/// @param splist
+/// @param para_dim
+/// @param dim
+/// @param degrees
+/// @param control_mesh_resolutions
+/// @param nthreads
+inline void RaiseMismatch(const CoreSplineVector& splist,
+                          const std::string name,
+                          const int para_dim,
+                          const int dim,
+                          const IntVector& degrees,
+                          const IntVector& control_mesh_resolutions,
+                          const int nthreads) {
+  // for verbose output
+  std::unordered_map<std::string, IntVectorVector> mismatches{};
+
+  // check flags
+  bool check_name{}, check_para_dim{}, check_dim{}, check_degrees{},
+      check_control_mesh_resolutions{};
+  // input vector sizes
+  const int d_size = degrees.size();
+  const int cmr_size = control_mesh_resolutions.size();
+  // true para_dim to use as reference -> will be updated / compared
+  int ref_para_dim{para_dim};
+
+  // parse input and allocate vector for mismatch book keeping
+  if (name.size() > 0) {
+    check_name = true;
+    mismatches["name"].resize(nthreads);
+  }
+  if (dim > 0) {
+    check_dim = true;
+    mismatches["dim"].resize(nthreads);
+  }
+  if (d_size > 0) {
+    check_degrees = true;
+    ref_para_dim = d_size;
+    mismatches["degrees"].resize(nthreads);
+  }
+  if (cmr_size > 0) {
```

Comment:
generally agree with kvs check, this will require a bit of changes and i'd say next PR, since the scope of this PR is to implement core multipatch that's compatible with previous implementations

Reply:
Next PR is fine for me